control/grpc.py

```diff
@@ -236,41 +266,173 @@ def __init__(self, config: GatewayConfig, gateway_state: GatewayStateHandler, rp
+        assert key_type, "Key type can't be empty"
+        assert key_value, "Key value can't be empty"
+
+        tmp_dir_names = self.get_directories_for_key_file(key_type, subsysnqn, create_dir = True)
+        if not tmp_dir_names:
+            return None
 
-        psk_dir = f"/tmp/psk/{subsysnqn}"
-        psk_file = f"{psk_dir}/{hostnqn}"
+        file = None
+        filepath = None
+        keyfile_prefix = f"{hostnqn}_"
         try:
-            os.makedirs(psk_dir, 0o755, True)
+            (file_fd, filepath) = tempfile.mkstemp(prefix=keyfile_prefix, dir=tmp_dir_names[0], text=True)
         except Exception:
-            self.logger.exception(f"Error creating directory {psk_dir}")
+            self.logger.exception("Error creating key file")
+            return None
+        if not filepath:
+            self.loger.error("Error creating key file")
             return None
         try:
-            with open(psk_file, 'wt') as f:
-                print(psk_value, end="", file=f)
-            os.chmod(psk_file, 0o600)
+            with open(file_fd, "wt") as f:
+                f.write(key_value)
         except Exception:
-            self.logger.exception(f"Error creating file {psk_file}")
+            self.logger.exception(f"Error creating file")
+            try:
+                os.remove(filepath)
+            except Exception:
+                pass
             return None
-        return psk_file
+        return filepath
+
+    def create_host_psk_file(self, subsysnqn : str, hostnqn : str, key_value : str) -> str:
+        return self.create_host_key_file(self.PSK_PREFIX, subsysnqn, hostnqn, key_value)
+
+    def create_host_dhchap_file(self, subsysnqn : str, hostnqn : str, key_value : str) -> str:
+        return self.create_host_key_file(self.DHCHAP_PREFIX, subsysnqn, hostnqn, key_value)
+
+    def remove_host_key_file(self, key_type : str, subsysnqn : str, hostnqn : str) -> None:
+        assert key_type, "Key type can't be empty"
+        assert subsysnqn, "Subsystem NQN can't be empty"
+
+        tmp_dir_names = self.get_directories_for_key_file(key_type, subsysnqn, create_dir = False)
+        if not tmp_dir_names:
+            return
+
+        for one_tmp_dir in tmp_dir_names:
+            if not hostnqn:
+                try:
+                    shutil.rmtree(one_tmp_dir, ignore_errors = True)
+                except Exception:
+                    pass
+                continue
+            key_files = []
+            try:
+                key_files = os.listdir(one_tmp_dir)
+            except Exception:
+                self.logger.exception(f"Error listing files in {one_tmp_dir}")
+                continue
+            if not key_files:
+                continue
+            for one_file in key_files:
+                if not one_file.startswith(f"{hostnqn}_"):
+                    continue
+                try:
+                    os.remove(one_tmp_dir + "/" + one_file)
```

Comment:
I'd the same thing here with `pathlib`.

Reply:
As I said, I preferred not to use pathlib here as the file deletions might affect the iteration. I didn't want to deal with it. And it's not that pathlib gives me such a benefit.